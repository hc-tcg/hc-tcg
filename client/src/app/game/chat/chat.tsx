import {SyntheticEvent, useState} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import classnames from 'classnames'
import {getChatMessages, getOpponentName, getPlayerStates} from 'logic/game/game-selectors'
import {chatMessage} from 'logic/game/game-actions'
import {getPlayerId} from 'logic/session/session-selectors'
import {getSettings} from 'logic/local-settings/local-settings-selectors'
import css from './chat.module.scss'
import Button from 'components/button'
import {setSetting} from 'logic/local-settings/local-settings-actions'
import {useDrag} from '@use-gesture/react'

function Chat() {
	const dispatch = useDispatch()
	const chatMessages = useSelector(getChatMessages)
	const playerStates = useSelector(getPlayerStates)
	const playerId = useSelector(getPlayerId)
	const settings = useSelector(getSettings)
	const opponent = useSelector(getOpponentName)
	const [chatPos, setChatPos] = useState({x: 0, y: 0})

	const bindChatPos = useDrag((params: any) => {
		setChatPos({
			x: params.offset[0],
			y: params.offset[1],
		})
	})

	if (settings.showChat !== 'on') return null

	const handleNewMessage = (ev: SyntheticEvent<HTMLFormElement>) => {
		ev.preventDefault()
		const form = ev.currentTarget
		const messageEl = form.message as HTMLInputElement
		const message = messageEl.value.trim()
		messageEl.value = ''
		messageEl.focus()
		if (message.length === 0) return
		dispatch(chatMessage(message))
	}

	const closeChat = () => {
		dispatch(setSetting('showChat', 'off'))
	}

	// @TODO: Repopulate chat messages after reconnecting
	return (
		<div
			className={css.chat}
			style={{top: chatPos.y, left: chatPos.x, width: '94vw', height: '50vh'}}
		>
			<div className={css.header} {...bindChatPos()}>
				<p>Chatting with {opponent}</p>
				<button onClick={closeChat}>
					<img src="/images/CloseX.svg" alt="close" />
				</button>
			</div>

			<div className={css.messagesWrapper}>
				<div className={css.messages}>
					{chatMessages
						.slice()
						.reverse()
						.map((msg) => {
							const time = new Date(msg.createdAt).toLocaleString()
							const hmTime = new Date(msg.createdAt).toLocaleTimeString([], {
								hour: '2-digit',
								minute: '2-digit',
							})
							const name = playerStates?.[msg.playerId]?.playerName || 'unknown'
							const isPlayer = playerId === msg.playerId
							return (
								<p
									key={msg.createdAt}
									className={classnames(css.message, {
										[css.player]: isPlayer,
										[css.opponent]: !isPlayer,
									})}
									title={time}
								>
									<span className={css.time}>{hmTime}</span>
									<span className={css.playerName}>{name}</span>
									<span className={css.text}>
										{settings.profanityFilter !== 'off' ? msg.censoredMessage : msg.message}
									</span>
								</p>
							)
						})}
				</div>
			</div>

			<form className={css.publisher} onSubmit={handleNewMessage}>
				<input autoComplete="off" autoFocus name="message" maxLength={140} />
				<Button variant="default" size="small">
					Send
				</Button>
			</form>
		</div>
	)
}

export default Chat
