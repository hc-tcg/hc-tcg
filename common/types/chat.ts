import {MessageTextT} from './game-state'

export type MessageInfoT = {
	createdAt: number
	message: Array<MessageTextT>
	playerId: string
	systemMessage: boolean
}
