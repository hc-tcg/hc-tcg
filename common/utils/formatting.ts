import ProfaneWords from '../config/profanity-seed.json'

/* Config object used by `formatedText`. */
export type Config = {
	/* If true, censor the text using the list common/config/profanity-seed.json. */
	censor?: boolean
	/* If true, allow $$ formatting to be used. */
	'enable-$'?: boolean
}

export type Format =
	| 'player'
	| 'opponent'
	| 'effect'
	| 'item'
	| 'attack'
	| 'good'
	| 'bad'
	| 'italic'
	| 'bold'
	| 'keyword'

export type FormattedTextNode =
	| ListNode
	| EmptyNode
	| TextNode
	| FormatNode
	| DifferentTextNode
	| ProfanityNode
	| EmojiNode
	| LineBreakNode
	| LineNode
	| TabNode

export type ListNode = {
	TYPE: 'ListNode'

	nodes: FormattedTextNode[]
}
export function ListNode(nodes: FormattedTextNode[]): ListNode {
	return {
		TYPE: 'ListNode',
		nodes,
	}
}

export type EmptyNode = {
	TYPE: 'EmptyNode'
}
export function EmptyNode(): EmptyNode {
	return {TYPE: 'EmptyNode'}
}

export type TextNode = {
	TYPE: 'TextNode'

	text: string
}
export function TextNode(text: string): TextNode {
	return {TYPE: 'TextNode', text}
}

export type FormatNode = {
	TYPE: 'FormatNode'

	format: Format
	text: FormattedTextNode
}
export function FormatNode(format: Format, text: FormattedTextNode): FormatNode {
	return {
		TYPE: 'FormatNode',
		format,
		text,
	}
}

const formatDict: Record<string, Format> = {
	p: 'player',
	o: 'opponent',
	e: 'effect',
	m: 'item',
	v: 'attack',
	g: 'good',
	b: 'bad',
	k: 'keyword',
}
export function formatNodefromShorthand(
	formatShorthand: string,
	text: FormattedTextNode
): FormatNode {
	let format = formatDict[formatShorthand]
	if (format == undefined) {
		throw new Error(`Format ${format} not found.`)
	}
	return {TYPE: 'FormatNode', format, text}
}

export type DifferentTextNode = {
	TYPE: 'DifferentTextNode'

	playerText: FormattedTextNode
	opponentText: FormattedTextNode
}
export function DifferentTextNode(
	playerText: FormattedTextNode,
	opponentText: FormattedTextNode
): DifferentTextNode {
	return {
		TYPE: 'DifferentTextNode',
		playerText,
		opponentText,
	}
}

export type EmojiNode = {
	TYPE: 'EmojiNode'

	emoji: string
}
export function EmojiNode(emoji: string): EmojiNode {
	return {
		TYPE: 'EmojiNode',
		emoji,
	}
}

export type ProfanityNode = {
	TYPE: 'ProfanityNode'

	text: string
}
export function ProfanityNode(text: string): ProfanityNode {
	return {
		TYPE: 'ProfanityNode',
		text,
	}
}

export function censorProfanityNode(node: ProfanityNode) {
	return '*'.repeat(node.text.length)
}

export type LineBreakNode = {
	TYPE: 'LineBreakNode'
}
export function LineBreakNode(): LineBreakNode {
	return {TYPE: 'LineBreakNode'}
}

export type LineNode = {
	TYPE: 'LineNode'
}
export function LineBreak(): LineNode {
	return {TYPE: 'LineNode'}
}

export type TabNode = {
	TYPE: 'TabNode'
}
export function TabNode(): TabNode {
	return {TYPE: 'TabNode'}
}

// The special characters that can end an expression.
const SPECIAL_CHARACTERS = [...'${}|*:\n\t']

const messageParseOptions: Array<
	[
		(text: string, config: Config) => boolean,
		(text: string, config: Config) => [FormattedTextNode, string],
	]
> = [
	[
		(text: string, config: Config) => {
			if (config['enable-$'] === undefined || config['enable-$'] === true) {
				return text.startsWith('$')
			}
			return false
		},
		(text: string, config: Config) => {
			// Expecting the format $fFormat Node$ where f is a format character
			let format = text[1]
			text = text.slice(2)

			let [node, remaining] = parseNodesUntil(
				text,
				(remaining) => remaining.startsWith('$'),
				config
			)

			if (node.TYPE === 'EmptyNode') {
				throw new Error('Expected an expression, not $')
			}

			if (remaining.length == 0 || remaining[0] != '$') {
				throw new Error('Expected $')
			}

			return [formatNodefromShorthand(format, node), remaining.slice(1)]
		},
	],
	[
		(text: string, _: Config) => text.startsWith('{'),
		(text: string, config: Config) => {
			// expecting the format {MesageTreeNode,|Node,}
			let remaining = text.slice(1)

			let firstNode
			;[firstNode, remaining] = parseSingleNode(remaining, config)

			if (remaining[0] !== '|') {
				throw new Error('Expected |')
			}

			remaining = remaining.slice(1)

			let secondNode
			;[secondNode, remaining] = parseSingleNode(remaining, config)

			if (remaining[0] !== '}') {
				throw new Error('Expected } to close expression.')
			}

			remaining = remaining.slice(1)

			return [DifferentTextNode(firstNode, secondNode), remaining]
		},
	],
	[
		(text: string, _: Config) => text.startsWith('**'),
		(text: string, config: Config) => {
			// There is no bold because the string isn't long enough.
			if (text.length == 2) {
				return parseTextNode(text, config)
			}

			// If there is no set of ** in the rest of the message, continue like this is regular text
			if (!includesNotEscaped(text.slice(2), '**')) {
				return parseTextNode(text, config)
			}

			// Otherwise lets parse a bold node list
			let [nodes, remaining] = parseNodesUntil(
				text.slice(2),
				(remaining) => remaining.startsWith('**'),
				config
			)
			remaining = remaining.slice(2)
			return [FormatNode('bold', nodes || EmptyNode()), remaining]
		},
	],
	[
		(text: string) => text.startsWith('*'),
		(text: string, config: Config) => {
			// There is no italic because the string isn't long enough.
			if (text.length == 1) {
				return parseTextNode(text, config)
			}

			// If there is no * in the rest of the message, continue like this is regular text
			if (!includesNotEscaped(text.slice(1), '*')) {
				return parseTextNode(text, config)
			}

			// Otherwise we parse a italic node list.
			let [nodes, remaining] = parseNodesUntil(
				text.slice(1),
				(remaining) => remaining.startsWith('*'),
				config
			)
			remaining = remaining.slice(1)
			return [FormatNode('italic', nodes || EmptyNode()), remaining]
		},
	],
	[
		(text: string, _: Config) => text.startsWith(':'),
		(text: string, _: Config) => {
			let remaining = text.slice(1)

			let emojiText: string
			;[emojiText, remaining] = parseUntil(remaining, [':'])

			if (remaining[0] !== ':') {
				throw new Error('Expected : to close expression.')
			}

			return [EmojiNode(emojiText), remaining.slice(1)]
		},
	],
	[
		(text: string, _: Config) => text.startsWith('\n'),
		(text: string, _: Config) => {
			return [LineBreakNode(), text.slice(1)]
		},
	],
	[
		(text: string, _: Config) => text.startsWith('\t'),
		(text: string, _: Config) => {
			return [TabNode(), text.slice(1)]
		},
	],
	[(_) => true, parseTextNode],
]

function includesNotEscaped(text: string, sequence: string): boolean {
	let index = text.indexOf(sequence)
	if (index == -1) {
		return false
	}

	if (index == 0) {
		return true
	}

	if (text[index - 1] == '\\') {
		return false
	}

	return true
}

function isAlphanumeric(char: string) {
	let charCode = char.charCodeAt(0)
	return (
		(charCode > 47 && charCode < 58) ||
		(charCode > 64 && charCode < 91) ||
		(charCode > 96 && charCode < 123)
	)
}

function createCensoredTextNodes(text: string): TextNode | ProfanityNode | ListNode {
	let nodes = []

	let lowercaseText = text.toLowerCase()

	for (const word of ProfaneWords) {
		while (true) {
			let startIndex = lowercaseText.indexOf(word)

			if (startIndex == -1) {
				break
			}

			let isSpaceBefore = true
			if (startIndex >= 1) {
				if (isAlphanumeric(text[startIndex - 1])) {
					isSpaceBefore = false
				}
			}

			let isSpaceAfter = true
			if (startIndex + word.length <= text.length - 1) {
				if (isAlphanumeric(text[startIndex + word.length])) {
					isSpaceAfter = false
				}
			}

			let textBefore = text.slice(0, startIndex)

			if (isSpaceBefore && isSpaceAfter) {
				if (textBefore.length > 0) {
					nodes.push(TextNode(textBefore))
				}
				nodes.push(ProfanityNode(text.slice(startIndex, startIndex + word.length)))
			} else {
				nodes.push(TextNode(text.slice(0, startIndex + word.length)))
			}

			text = text.slice(startIndex + word.length)
			lowercaseText = lowercaseText.slice(startIndex + word.length)
		}
	}

	if (nodes.length != 0) {
		if (text.length !== 0) {
			nodes.push(TextNode(text))
		}
		if (nodes.length === 1) {
			return nodes[0]
		}
		return ListNode(nodes)
	}

	return TextNode(text)
}

/* Parse the raw text that is part of a text mode or emoji node, handling escape sequences. */
function parseUntil(text: string, until: Array<string>): [string, string] {
	// We take characters until we get to something that is probably a parser
	let out = ''
	let i = 0

	let isEscaped = text[0] == '\\'
	let nextChar: string | undefined = text[0]

	while (true) {
		if (!isEscaped) {
			out += nextChar
		}
		i++

		if (i >= text.length) {
			break
		}
		nextChar = text.at(i)
		if (nextChar == undefined) {
			break
		}

		if (!isEscaped && until.includes(nextChar)) {
			break
		}

		isEscaped = nextChar === '\\'
	}

	if (text[text.length - 1] == '\\') {
		out += '\\'
	}

	return [out, text.slice(i)]
}

// Parse nodes until a predicate matches, or there is an error
function parseNodesWhile(
	text: string,
	matches: (remaining: string) => boolean,
	config: Config
): [FormattedTextNode, string] {
	let remaining = text
	let nodes: FormattedTextNode[] = []

	try {
		while (true) {
			if (!matches(remaining)) {
				break
			}

			if (remaining.length === 0) {
				throw new Error('Ran out of text when parsing (Unexpected EOF).')
			}

			let node
			;[node, remaining] = parseSingleNode(remaining, config)
			nodes.push(node)
		}
	} catch (e) {
		if (remaining.length != 0) {
			nodes.push(TextNode(remaining))
			remaining = ''
		}
	}

	let formatNode: FormattedTextNode
	if (nodes.length == 0) {
		formatNode = EmptyNode()
	} else if (nodes.length == 1) {
		formatNode = nodes[0]
	} else {
		formatNode = ListNode(nodes)
	}

	return [formatNode, remaining]
}

function parseNodesUntil(
	text: string,
	matches: (remaining: string) => boolean,
	config: Config
): [FormattedTextNode, string] {
	return parseNodesWhile(text, (remaining) => !matches(remaining), config)
}

/* Parse all Nodes until the end of the string. */
function parseNodesUntilEmpty(text: string, config: Config): FormattedTextNode {
	let [nodes, _] = parseNodesWhile(text, (remaining) => remaining.length >= 1, config)
	return nodes
}

/* Parse a TextNode */
function parseTextNode(text: string, config: Config): [FormattedTextNode, string] {
	let remaining
	;[text, remaining] = parseUntil(text, SPECIAL_CHARACTERS)

	let textNodes: FormattedTextNode
	if (config.censor) {
		textNodes = createCensoredTextNodes(text)
	} else {
		textNodes = TextNode(text)
	}

	return [textNodes, remaining]
}

/* Parse text into a single node */
function parseSingleNode(text: string, config: Config): [FormattedTextNode, string] {
	for (let [condition, parser] of messageParseOptions) {
		if (condition(text, config)) {
			return parser(text, config)
		}
	}
	throw new Error(`No matching parser found for \`${text}\``)
}

/**
 * Format a text
 *
 * Guide to symbols
 * Player A - Player that generated the log | Player B - other player
 * {A|B} A shows to player A, B shows to player B
 * $p Player
 * $o Opponent
 * $e Effect card
 * $m Item card
 * $v Hermit attack
 * $g Good (healing, heads)
 * $b Bad (damage, tails)
 * $i Image
 * $k Keyword
 *
 * All symbols can be escaped with backslash.
 *
 * @param text - The text to format
 * @param config - The config as a dict.
 * @return The formatted text.
 */
export function formatText(text: string, config?: Config): FormattedTextNode {
	config = config || {}

	try {
		return parseNodesUntilEmpty(text, config)
	} catch (e) {
		return TextNode('There was a unrecoverable formatting error')
	}
}

/* Censor a string using `common/config/profanity-seed.json`. */
export function censorString(text: string) {
	let node = createCensoredTextNodes(text)

	if (node.TYPE === 'TextNode') {
		return node.text
	} else if (node.TYPE === 'ProfanityNode') {
		return censorProfanityNode(node)
	}

	let outputText = []

	let listNode = node as ListNode
	for (let textNode of listNode.nodes) {
		if (textNode.TYPE === 'TextNode') {
			outputText.push(textNode.text)
		} else if (textNode.TYPE === 'ProfanityNode') {
			outputText.push(censorProfanityNode(textNode))
		}
	}

	return outputText.join('')
}

/* Concat a list of formatted `FormattedTextNode` into a single `FormattedTextNode */
export function concatFormattedTextNodes(...nodes: Array<FormattedTextNode>): ListNode {
	return ListNode(nodes)
}
