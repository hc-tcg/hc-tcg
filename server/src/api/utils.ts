import url from 'url'
import {Request} from 'express'

export function requestUrlRoot(req: Request) {
	return url.format({
		protocol: req.protocol,
		host: req.get('host'),
	})
}

export function joinUrl(a: string, b: string) {
	if (a.endsWith('/')) {
		a = a.slice(0, a.length - 1)
	}
	if (b.startsWith('/')) {
		b = b.slice(1, b.length)
	}
	return `${a}/${b}`
}
