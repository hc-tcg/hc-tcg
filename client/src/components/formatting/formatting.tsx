import classNames from 'classnames'
import {FormattedTextNode, censorProfanityNode} from 'common/utils/formatting'
import css from './formatting.module.scss'

type DisplaySettings = {
	isSelectable?: boolean
	isOpponent?: boolean
	censorProfanity?: boolean
}

function nodeToHtml(node: FormattedTextNode, settings: DisplaySettings) {
	let textCssClasses = []

	if (settings.isSelectable != false) {
		textCssClasses.push(css['selectable'])
	}

	switch (node.TYPE) {
		case 'ListNode':
			let html = []

			for (let child of node.nodes) {
				html.push(nodeToHtml(child, settings))
			}
			return <span>{html}</span>
		case 'EmptyNode':
			return <span />
		case 'PlaintextNode':
			return <span className={classNames(...textCssClasses)}>{node.text}</span>
		case 'FormatNode':
			return (
				<span
					className={classNames(
						css[node.format],
						settings.isOpponent ? css.viewedByOpponent : '',
						...textCssClasses,
					)}
				>
					{nodeToHtml(node.text, settings)}
				</span>
			)
		case 'DifferentTextNode':
			return (
				<span className={classNames(...textCssClasses)}>
					{settings.isOpponent
						? nodeToHtml(node.opponentText, settings)
						: nodeToHtml(node.playerText, settings)}
				</span>
			)
		case 'ProfanityNode':
			if (settings.censorProfanity) {
				return (
					<span className={classNames(...textCssClasses)}>
						{' '}
						{censorProfanityNode(node)}{' '}
					</span>
				)
			}
			return <span className={classNames(...textCssClasses)}>{node.text}</span>
		case 'EmojiNode':
			const link = `/images/hermits-emoji/${node.emoji.toLowerCase()}.png`
			const alt = `:${node.emoji}:`

			return (
				<img
					className={classNames(css.emoji, ...textCssClasses)}
					src={link}
					alt={alt}
				/>
			)
		case 'LineBreakNode':
			return <br />
		case 'TabNode':
			return <span className={css.tab}></span>
		case 'LineNode':
			return <span className={css.line} />
	}
}

export const FormattedText = (
	text: FormattedTextNode | undefined,
	settings?: DisplaySettings,
) => {
	settings = settings || {}
	if (!text) return <span />

	return nodeToHtml(text, settings)
}
