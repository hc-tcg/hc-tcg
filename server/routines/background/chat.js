import {takeEvery} from 'redux-saga/effects'
import {broadcast} from '../../utils/comm'
import profanityFilter from '../../utils/profanity'
import {GameModel} from '../../models/game-model'

/**
 * @typedef {import("redux").AnyAction} AnyAction
 * @typedef {import("redux-saga").SagaIterator} SagaIterator
 */

/**
 * @param {string} type
 * @param {GameModel} game
 */
const gameAction = (type, game) => (action) => {
	return action.type === type && !!game.players[action.playerId]
}

/**
 * @param {GameModel} game
 * @param {AnyAction} action
 * @returns {SagaIterator}
 */
function* chatMessageSaga(game, action) {
	const {payload: message, playerId} = action
	if (typeof message !== 'string') return
	if (message.length < 1) return
	if (message.length > 140) return
	game.chat.push({
		createdAt: Date.now(),
		message,
		censoredMessage: profanityFilter(message),
		playerId,
	})
	broadcast(game.getPlayers(), 'CHAT_UPDATE', game.chat)
}

/**
 * @param {GameModel} game
 * @returns {SagaIterator}
 */
function* chatSaga(game) {
	yield takeEvery(gameAction('CHAT_MESSAGE', game), chatMessageSaga, game)
}

export default chatSaga
