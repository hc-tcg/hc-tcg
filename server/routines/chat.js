import {takeEvery} from 'redux-saga/effects'
import {broadcast} from '../utils/socket'

/**
 * @typedef {import("../classes/game").Game} Game
 */

/**
 * @param {string} type
 * @param {Game} game
 */
const gameAction = (type, game) => (action) => {
	return action.type === type && !!game.players[action.playerId]
}

/**
 * @param {Game} game
 */
function* chatSaga(game) {
	yield takeEvery(gameAction('CHAT_MESSAGE', game), function* (action) {
		const {payload: message, playerId} = action
		if (typeof message !== 'string') return
		if (message.length < 1) return
		if (message.length > 140) return
		game.chat.push({
			createdAt: Date.now(),
			message,
			playerId,
		})
		broadcast(game.getPlayerValues(), 'CHAT_UPDATE', game.chat)
	})
}

export default chatSaga
