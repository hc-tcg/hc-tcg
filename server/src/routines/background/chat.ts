import {takeEvery} from 'redux-saga/effects'
import {broadcast} from '../../utils/comm'
import profanityFilter from '../../../../common/utils/profanity'
import {PlayerModel} from 'common/models/player-model'
import {GameModel} from 'common/models/game-model'
import {AnyAction} from 'redux'
import {SagaIterator} from 'redux-saga'

const gameAction =
	(type: string, game: {players: Record<string, PlayerModel>}) => (action: any) => {
		return action.type === type && !!game.players[action.playerId]
	}

function* chatMessageSaga(game: GameModel, action: AnyAction): SagaIterator {
	const {payload: message, playerId} = action
	if (typeof message !== 'string') return
	if (message.length < 1) return
	if (message.length > 140) return
	game.chat.push({
		createdAt: Date.now(),
		message,
		censoredMessage: profanityFilter(message),
		playerId,
	})
	broadcast(game.getPlayers(), 'CHAT_UPDATE', game.chat)
}

function* chatSaga(game: GameModel): SagaIterator {
	yield takeEvery(gameAction('CHAT_MESSAGE', game), chatMessageSaga, game)
}

export default chatSaga
