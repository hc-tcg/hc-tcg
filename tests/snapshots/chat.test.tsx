import {describe, expect, test} from '@jest/globals'
import * as ReactTestRenderer from 'react-test-renderer'
import {ChatContent, ChatMessageDisplay} from 'client/app/game/chat/chat'
import {formatText, LineNode} from 'common/utils/formatting'

describe('Test chat messages', () => {
	test('Basic Test', () => {
		let chatMessages: Array<ChatMessageDisplay> = [
			{
				message: LineNode(),
				isBattleLogMessage: true,
				sender: 'playerOne',
				createdAt: 0,
			},
			{
				message: formatText('Hello Everyone, I am Player One.'),
				isBattleLogMessage: false,
				sender: 'playerOne',
				createdAt: 0,
			},
			{
				message: formatText('Hello Everyone, I am Player Two.'),
				isBattleLogMessage: false,
				sender: 'playerTwo',
				createdAt: 0,
			},
			{
				message: LineNode(),
				isBattleLogMessage: true,
				sender: 'playerTwo',
				createdAt: 0,
			},
		]

		expect(
			ReactTestRenderer.create(
				<ChatContent
					chatMessages={chatMessages}
					showLog={true}
					profanityFilterEnabled={true}
					isSpectating={false}
					playerNames={['Player One', 'Player Two']}
				/>,
			).toJSON(),
		).toMatchSnapshot()
	})
})
