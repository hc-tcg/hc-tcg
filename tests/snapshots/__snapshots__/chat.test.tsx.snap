// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test chat messages Basic Test 1`] = `
<div>
  <div>
    <div>
      <span>
        YOUR
         TURN
      </span>
      <span />
    </div>
    <div>
      <span>
        07:00 PM
      </span>
      <span
        className=""
      >
        <span
          className=""
        >
          Hello Everyone, I am Player One.
        </span>
      </span>
    </div>
    <div>
      <span>
        07:00 PM
      </span>
      <span
        className=""
      >
        <span
          className=""
        >
          Hello Everyone, I am Player Two.
        </span>
      </span>
    </div>
    <div>
      <span>
        PLAYER TWO
         TURN
      </span>
      <span />
    </div>
  </div>
</div>
`;
